src/lua/msgpack.c

```diff
@@ -809,6 +811,190 @@ luamp_object_iterator(struct lua_State *L)
+ */
+static int
+luamp_object_get_integer_array(struct lua_State *L, struct luamp_object *obj,
+			       uint64_t idx, bool is_signed)
+{
+	if (is_signed)
+		return 0;
+
+	const char *it = obj->data;
+	uint32_t size = mp_decode_array(&it);
+	if (idx == 0 || idx > size) {
+		return 0;
+	}
+	for (; idx > 1; --idx)
+		mp_next(&it);
+	luamp_decode(L, obj->cfg, &it);
+	return 1;
+}
+
+/**
+ * Helper for indexing `msgpack.object` in the case, when the indexation key
+ * is an integer. The underlying MsgPack is expected to be a map.
+ */
+static int
+luamp_object_get_integer_map(struct lua_State *L, struct luamp_object *obj,
+			     uint64_t key, bool is_signed)
+{
+	const char *it = obj->data;
+	uint32_t size = mp_decode_map(&it);
+	for (uint32_t i = 0; i < size; ++i, mp_next(&it)) {
+		switch (mp_typeof(*it)) {
+		case MP_UINT: {
+			uint64_t probe = mp_decode_uint(&it);
+			if (!is_signed && key == probe) {
+				luamp_decode(L, obj->cfg, &it);
+				return 1;
+			}
+			break;
+		}
+		case MP_INT: {
+			int64_t probe = mp_decode_int(&it);
+			if (is_signed) {
+				if ((int64_t)key == probe) {
+					luamp_decode(L, obj->cfg, &it);
+					return 1;
+				}
+			} else {
+				if (probe >= 0 && key == (uint64_t)probe) {
+					luamp_decode(L, obj->cfg, &it);
+					return 1;
+				}
+			}
+			break;
+		}
+		default:
+			mp_next(&it);
+		}
+	}
+	return 0;
+}
+
+/**
+ * Helper for indexing `msgpack.object` in the case, when the indexation key
+ * is an integer.
+ */
+static int
+luamp_object_get_integer(struct lua_State *L, struct luamp_object *obj,
+			 uint64_t num, bool is_signed)
+{
+	switch (mp_typeof(*obj->data)) {
+	case MP_MAP:
+		return luamp_object_get_integer_map(L, obj, num, is_signed);
+	case MP_ARRAY:
+		return luamp_object_get_integer_array(L, obj, num, is_signed);
+	default:
+		unreachable();
+	}
+}
+
+/**
+ * Helper for indexing `msgpack.object` in the case, when the indexation key
+ * is a string. The underlying MsgPack is expected to be a map.
+ */
+static int
+luamp_object_get_string(struct lua_State *L, struct luamp_object *obj,
+			const char *key, size_t key_len)
+{
+	const char *it = obj->data;
+	uint32_t size = mp_decode_map(&it);
+	for (uint32_t i = 0; i < size; ++i, mp_next(&it)) {
+		if (mp_typeof(*it) != MP_STR) {
+			mp_next(&it);
+			continue;
+		}
+
+		uint32_t probe_len;
+		const char *probe = mp_decode_str(&it, &probe_len);
+		if (key_len == probe_len && strncmp(key, probe, key_len) == 0) {
+			luamp_decode(L, obj->cfg, &it);
+			return 1;
+		}
+	}
+	return 0;
+}
+
+/**
+ * Takes a `msgpack.object` and an indexation key as the arguments, indexes
+ * the MsgPack stored in the `msgpack.object` and pushes the result to Lua stack
+ * or throws an error.
+ */
+static int
+luamp_object_get(struct lua_State *L)
+{
+	struct luamp_object *obj = luamp_check_object(L, 1);
+
+	enum mp_type type = mp_typeof(*obj->data);
+	if (type != MP_MAP && type != MP_ARRAY)
+		return luaL_error(L, "not an array or map");
+
+	switch (lua_type(L, 2)) {
```

Comment:
lua_rawget(L, -2) + catch error &  return nil if error

Reply:
`lua_rawget` does not throw errors. 

Rewrote the patch: now we do `lua_rawget` instead of explicit indexing.